atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -140,15 +147,75 @@
 import com.palantir.timestamp.TimestampService;
 
 @SuppressWarnings("checkstyle:all")
+@RunWith(Parameterized.class)
 public class SnapshotTransactionTest extends AtlasDbTestCase {
+    private static final String SYNC = "sync";
+    private static final String ASYNC = "async";
+    private final String name;
+    private final Function<Transaction, Transaction> transactionWrapper;
+    private final Map<String, ExpectationFactory<KeyValueService, Cell, Long, LockService, Expectations>>
+            expectationsMapping =
+            ImmutableMap.<String, ExpectationFactory<KeyValueService, Cell, Long, LockService, Expectations>>builder()
+                    .put(SYNC, SnapshotTransactionTest.this::syncGetExpectation)
+                    .put(ASYNC, SnapshotTransactionTest.this::asyncGetExpectation)
+                    .build();
+
+    @FunctionalInterface
+    interface ExpectationFactory<One, Two, Three, Four, Five> {
+        Five apply(One one, Two two, Three three, Four four) throws Exception;
```

Comment:
Hmm, I'm not sure how/if we plan on reusing this, but currently this class represents a `QuadFunction` rather than an `ExpectationFactory`. I think if you want to reuse this name, `Five` should be locked to `Expectations`. Even then it seems like you might as well just set `One` through `Four` to be the types they are actually set to, and not have the type params?